In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
visa_df=pd.read_csv("Visadataset.csv")
visa_df.head()

case_id continent education_of_employee has_job_experience  \
0  EZYV01      Asia           High School                  N   
1  EZYV02      Asia              Master's                  Y   
2  EZYV03      Asia            Bachelor's                  N   
3  EZYV04      Asia            Bachelor's                  N   
4  EZYV05    Africa              Master's                  Y   

  requires_job_training  no_of_employees  yr_of_estab region_of_employment  \
0                     N            14513         2007                 West   
1                     N             2412         2002            Northeast   
2                     Y            44444         2008                 West   
3                     N               98         1897                 West   
4                     N             1082         2005                South   

   prevailing_wage unit_of_wage full_time_position case_status  
0         592.2029         Hour                  Y      Denied  
1       83425.6500         Year                  Y   Certified  
2      122996.8600         Year                  Y      Denied  
3       83434.0300         Year                  Y      Denied  
4      149907.3900         Year                  Y   Certified

In [3]:
cat_columns=visa_df.select_dtypes(include='object').columns
num_columns=visa_df.select_dtypes(exclude='object').columns

**Encoding**

- Encoding means convert categorical columns data to Numerical data

- Because ML models expect the data should be in numerical format

- ML models developed by math algorithms

- It is very very important to apply encoding

- The following methods are

  - map 

  - np.where

  - one hot encoding

  - Label encoder

**Map**

- map is one method to convert categorical values to numerical values

- Take one categorical column
 
- Get the unique labels first

- make a dictionary with assign a number to each label

- for example case_status has two labels

  - Certified
 
  - Denied
 
- Assign 0 to certified, 1 to denied

- create a dictionary lebels as keys, numbers as values only

- {'Certified':0,'Denied':1}

In [ ]:
#step-1: read the column
#step-2: get the unique labels
#step-3: make a dictionary
#step-4: apply the map

In [18]:
visa_df['case_status'].unique()

array(['Denied', 'Certified'], dtype=object)

In [19]:
d={'Certified':0,'Denied':1}
visa_df['case_status_new']=visa_df['case_status'].map(d)
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

      case_status  case_status_new  
0          Denied                1  
1       Certified                0  
2          Denied                1  
3          Denied                1  
4       Certified                0  
...           ...              ...  
25475   Certified                0  
25476   Certified                0  
25477   Certified                0  
25478   Certified                0  
25479   Certified                0  

[25480 rows x 13 columns]

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
visa_df=pd.read_csv("Visadataset.csv")
visa_df.head()

case_id continent education_of_employee has_job_experience  \
0  EZYV01      Asia           High School                  N   
1  EZYV02      Asia              Master's                  Y   
2  EZYV03      Asia            Bachelor's                  N   
3  EZYV04      Asia            Bachelor's                  N   
4  EZYV05    Africa              Master's                  Y   

  requires_job_training  no_of_employees  yr_of_estab region_of_employment  \
0                     N            14513         2007                 West   
1                     N             2412         2002            Northeast   
2                     Y            44444         2008                 West   
3                     N               98         1897                 West   
4                     N             1082         2005                South   

   prevailing_wage unit_of_wage full_time_position case_status  
0         592.2029         Hour                  Y      Denied  
1       83425.6500         Year                  Y   Certified  
2      122996.8600         Year                  Y      Denied  
3       83434.0300         Year                  Y      Denied  
4      149907.3900         Year                  Y   Certified

In [3]:
cat_columns=visa_df.select_dtypes(include='object').columns
num_columns=visa_df.select_dtypes(exclude='object').columns

In [ ]:
# I want to apply, the map method for all categorical columns
# step-1: we need to get unique labels
# step-2: we need to get a numbers=len unique labels
# step-3: we need to create a dictionary

In [7]:
visa_df=pd.read_csv("Visadataset.csv")
unique_labels=sorted(visa_df['case_status'].unique())
list1=[i for i in range(len(unique_labels))]
dict1=dict(zip(unique_labels,list1))
dict1
visa_df['case_status_new']=visa_df['case_status'].map(dict1)
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

      case_status  case_status_new  
0          Denied                1  
1       Certified                0  
2          Denied                1  
3          Denied                1  
4       Certified                0  
...           ...              ...  
25475   Certified                0  
25476   Certified                0  
25477   Certified                0  
25478   Certified                0  
25479   Certified                0  

[25480 rows x 13 columns]

In [10]:
{(key,value) for key,value in zip(unique_labels,list1)}

{('Certified', 0), ('Denied', 1)}

In [13]:
visa_df=pd.read_csv("Visadataset.csv")
for j in cat_columns:
    unique_labels=sorted(visa_df[j].unique())
    list1=[i for i in range(len(unique_labels))]
    dict1=dict(zip(unique_labels,list1))
    dict1
    visa_df[f'{j}_new']=visa_df[j].map(dict1)
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage  ... case_status  \
0                     West         592.2029         Hour  ...      Denied   
1                Northeast       83425.6500         Year  ...   Certified   
2                     West      122996.8600         Year  ...      Denied   
3                     West       83434.0300         Year  ...      Denied   
4                    South      149907.3900         Year  ...   Certified   
...                    ...              ...          ...  ...         ...   
25475                South       77092.5700         Year  ...   Certified   
25476            Northeast      279174.7900         Year  ...   Certified   
25477                South      146298.8500         Year  ...   Certified   
25478                 West       86154.7700         Year  ...   Certified   
25479              Midwest       70876.9100         Year  ...   Certified   

      case_id_new  continent_new  education_of_employee_new  \
0               0              1                          2   
1               1              1                          3   
2               2              1                          0   
3               3              1                          0   
4               4              0                          3   
...           ...            ...                        ...   
25475       17204              1                          0   
25476       17205              1                          2   
25477       17206              1                          3   
25478       17207              1                          3   
25479       17209              1                          0   

       has_job_experience_new  requires_job_training_new  \
0                           0                          0   
1                           1                          0   
2                           0                          1   
3                           0                          0   
4                           1                          0   
...                       ...                        ...   
25475                       1                          1   
25476                       1                          0   
25477                       1                          0   
25478                       1                          1   
25479                       1                          0   

       region_of_emp

In [14]:
visa_df=pd.read_csv("Visadataset.csv")
for i in cat_columns:
    unique_labels=sorted(visa_df[i].unique())
    list1=[i for i in range(len(unique_labels))]
    dict1=dict(zip(unique_labels,list1))
    dict1
    visa_df[i]=visa_df[i].map(dict1)
visa_df

case_id  continent  education_of_employee  has_job_experience  \
0            0          1                      2                   0   
1            1          1                      3                   1   
2            2          1                      0                   0   
3            3          1                      0                   0   
4            4          0                      3                   1   
...        ...        ...                    ...                 ...   
25475    17204          1                      0                   1   
25476    17205          1                      2                   1   
25477    17206          1                      3                   1   
25478    17207          1                      3                   1   
25479    17209          1                      0                   1   

       requires_job_training  no_of_employees  yr_of_estab  \
0                          0            14513         2007   
1                          0             2412         2002   
2                          1            44444         2008   
3                          0               98         1897   
4                          0             1082         2005   
...                      ...              ...          ...   
25475                      1             2601         2008   
25476                      0             3274         2006   
25477                      0             1121         1910   
25478                      1             1918         1887   
25479                      0             3195         1960   

       region_of_employment  prevailing_wage  unit_of_wage  \
0                         4         592.2029             0   
1                         2       83425.6500             3   
2                         4      122996.8600             3   
3                         4       83434.0300             3   
4                         3      149907.3900             3   
...                     ...              ...           ...   
25475                     3       77092.5700             3   
25476                     2      279174.7900             3   
25477                     3      146298.8500             3   
25478                     4       86154.7700             3   
25479                     1       70876.9100             3   

       full_time_position  case_status  
0                       1            1  
1                       1            0  
2                       1            1  
3                       1            1  
4                       1            0  
...                   ...          ...  
25475                   1            0  
25476                   1            0  
25477                   0            0  
25478                   1            0  
25479                   1            0  

[25480 rows x 12 columns]

**Label Encoder**

- Label Encoder is a method to convert categorical data to numerical data

- Label Encoder is also same way, how we did in the above

- The for loop map method we explored

- But the label encoder also do the same

- It is under **sickit-learn** we usually called as **sklearn**

- under sklearn we have class called as preprocessing

- Because currently we are doing data preprocessing only

- Under preprocessing class we have a method called as **Label Encoder**

- sklearn

    - preprocessing

      - Label Encoder

- Any sklearn package has 3 steps

    - step-1: Read the package

    - step-2: Save the package

    - step-3: Apply fit transform on data

In [15]:
# Read the data
visa_df=pd.read_csv("Visadataset.csv")

In [17]:
# step-1: Read the package
from sklearn.preprocessing import LabelEncoder

# step-2: Save the package
le=LabelEncoder()

# step-3: Apply fit transform on data
visa_df['case_status']=le.fit_transform(visa_df['case_status'])

visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

       case_status  
0                1  
1                0  
2                1  
3                1  
4                0  
...            ...  
25475            0  
25476            0  
25477            0  
25478            0  
25479            0  

[25480 rows x 12 columns]

In [19]:
visa_df=pd.read_csv("Visadataset.csv")
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()
for i in cat_columns:
    visa_df[i]=le.fit_transform(visa_df[i])
visa_df

case_id  continent  education_of_employee  has_job_experience  \
0            0          1                      2                   0   
1            1          1                      3                   1   
2            2          1                      0                   0   
3            3          1                      0                   0   
4            4          0                      3                   1   
...        ...        ...                    ...                 ...   
25475    17204          1                      0                   1   
25476    17205          1                      2                   1   
25477    17206          1                      3                   1   
25478    17207          1                      3                   1   
25479    17209          1                      0                   1   

       requires_job_training  no_of_employees  yr_of_estab  \
0                          0            14513         2007   
1                          0             2412         2002   
2                          1            44444         2008   
3                          0               98         1897   
4                          0             1082         2005   
...                      ...              ...          ...   
25475                      1             2601         2008   
25476                      0             3274         2006   
25477                      0             1121         1910   
25478                      1             1918         1887   
25479                      0             3195         1960   

       region_of_employment  prevailing_wage  unit_of_wage  \
0                         4         592.2029             0   
1                         2       83425.6500             3   
2                         4      122996.8600             3   
3                         4       83434.0300             3   
4                         3      149907.3900             3   
...                     ...              ...           ...   
25475                     3       77092.5700             3   
25476                     2      279174.7900             3   
25477                     3      146298.8500             3   
25478                     4       86154.7700             3   
25479                     1       70876.9100             3   

       full_time_position  case_status  
0                       1            1  
1                       1            0  
2                       1            1  
3                       1            1  
4                       1            0  
...                   ...          ...  
25475                   1            0  
25476                   1            0  
25477                   0            0  
25478                   1            0  
25479                   1            0  

[25480 rows x 12 columns]

**fit and transform**

- fit means develop the logic

- transform means process the logic

- if we observe in map method first we developed dictionary logic

- then we apply that dictionary to column

- Developing the dictionary is kind of fit

- apply the dictionary to column means we are transforming the data cat to num

- whenever we are transforming or changing the data then use **fit transform**

- if we are not changing the data then only **fit**

**np.where**

- using np.where also we can change the categorical data to numerical

- But np.where is only for binary condition

- np.where is same as if-else

- if condition is True: will apply True value

- if condition is False: will apply False value

- so np.where use for only binary labels

- A column should have only two unique labels

In [22]:
#value=='certified' replace with 0
#otherwise 1
#np.where(con,0,1)
visa_df=pd.read_csv("Visadataset.csv")
con=visa_df['case_status']=='Certified'
visa_df['case_status']=np.where(con,0,1)
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

       case_status  
0                1  
1                0  
2                1  
3                1  
4                0  
...            ...  
25475            0  
25476            0  
25477            0  
25478            0  
25479            0  

[25480 rows x 12 columns]

**One hot encoder**

- One hot encoder if one will be ON, another will be OFF

- ON represents with 1

- OFF represents with 0

- For example case status has two unique labels

  - Certified

  - Denied

- One hot encoder creates new columns which is equal to number of unique labels

- For example: case status will create two new extra columns

  - case_status_certified

  - case_status_denied

|case_status|case_status_certified|case_status_denied|
|-|-|-|
|Denied|0|1|
|Certified|1|0|
|Denied|0|1|

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
visa_df=pd.read_csv("Visadataset.csv")
visa_df.head()

case_id continent education_of_employee has_job_experience  \
0  EZYV01      Asia           High School                  N   
1  EZYV02      Asia              Master's                  Y   
2  EZYV03      Asia            Bachelor's                  N   
3  EZYV04      Asia            Bachelor's                  N   
4  EZYV05    Africa              Master's                  Y   

  requires_job_training  no_of_employees  yr_of_estab region_of_employment  \
0                     N            14513         2007                 West   
1                     N             2412         2002            Northeast   
2                     Y            44444         2008                 West   
3                     N               98         1897                 West   
4                     N             1082         2005                South   

   prevailing_wage unit_of_wage full_time_position case_status  
0         592.2029         Hour                  Y      Denied  
1       83425.6500         Year                  Y   Certified  
2      122996.8600         Year                  Y      Denied  
3       83434.0300         Year                  Y      Denied  
4      149907.3900         Year                  Y   Certified

In [4]:
cat_columns=visa_df.select_dtypes(include='object').columns
num_columns=visa_df.select_dtypes(exclude='object').columns

**pd.get_dummies**

In [4]:
df=pd.get_dummies(visa_df['case_status'],dtype='int')
df['case_status']=visa_df['case_status'].values
df

Certified  Denied case_status
0              0       1      Denied
1              1       0   Certified
2              0       1      Denied
3              0       1      Denied
4              1       0   Certified
...          ...     ...         ...
25475          1       0   Certified
25476          1       0   Certified
25477          1       0   Certified
25478          1       0   Certified
25479          1       0   Certified

[25480 rows x 3 columns]

In [5]:
df=pd.get_dummies(visa_df['case_status'],dtype='int')
df['case_status']=visa_df['case_status'].values
df['dummy']=[i for i in range(len(visa_df))]
df

Certified  Denied case_status  dummy
0              0       1      Denied      0
1              1       0   Certified      1
2              0       1      Denied      2
3              0       1      Denied      3
4              1       0   Certified      4
...          ...     ...         ...    ...
25475          1       0   Certified  25475
25476          1       0   Certified  25476
25477          1       0   Certified  25477
25478          1       0   Certified  25478
25479          1       0   Certified  25479

[25480 rows x 4 columns]

In [6]:
df1=pd.get_dummies(visa_df['case_status'],dtype='int')
df1

Certified  Denied
0              0       1
1              1       0
2              0       1
3              0       1
4              1       0
...          ...     ...
25475          1       0
25476          1       0
25477          1       0
25478          1       0
25479          1       0

[25480 rows x 2 columns]

In [7]:
df2=visa_df[['case_status']]
df2

case_status
0          Denied
1       Certified
2          Denied
3          Denied
4       Certified
...           ...
25475   Certified
25476   Certified
25477   Certified
25478   Certified
25479   Certified

[25480 rows x 1 columns]

**Adding the two dataframes**

- Merge

- join

- concat

- append

In [15]:
pd.merge([df1,df2],on= "case_status",how="right")

TypeError: merge() missing 1 required positional argument: 'right'

In [23]:
df1.join(df2)

Certified  Denied case_status
0              0       1      Denied
1              1       0   Certified
2              0       1      Denied
3              0       1      Denied
4              1       0   Certified
...          ...     ...         ...
25475          1       0   Certified
25476          1       0   Certified
25477          1       0   Certified
25478          1       0   Certified
25479          1       0   Certified

[25480 rows x 3 columns]

In [24]:
pd.concat([df1,df2],axis=1)

Certified  Denied case_status
0              0       1      Denied
1              1       0   Certified
2              0       1      Denied
3              0       1      Denied
4              1       0   Certified
...          ...     ...         ...
25475          1       0   Certified
25476          1       0   Certified
25477          1       0   Certified
25478          1       0   Certified
25479          1       0   Certified

[25480 rows x 3 columns]

In [22]:
df1.append(df2)

AttributeError: 'DataFrame' object has no attribute 'append'

In [12]:
pd.get_dummies(visa_df,dtype='int')

no_of_employees  yr_of_estab  prevailing_wage  case_id_EZYV01  \
0                14513         2007         592.2029               1   
1                 2412         2002       83425.6500               0   
2                44444         2008      122996.8600               0   
3                   98         1897       83434.0300               0   
4                 1082         2005      149907.3900               0   
...                ...          ...              ...             ...   
25475             2601         2008       77092.5700               0   
25476             3274         2006      279174.7900               0   
25477             1121         1910      146298.8500               0   
25478             1918         1887       86154.7700               0   
25479             3195         1960       70876.9100               0   

       case_id_EZYV02  case_id_EZYV03  case_id_EZYV04  case_id_EZYV05  \
0                   0               0               0               0   
1                   1               0               0               0   
2                   0               1               0               0   
3                   0               0               1               0   
4                   0               0               0               1   
...               ...             ...             ...             ...   
25475               0               0               0               0   
25476               0               0               0               0   
25477               0               0               0               0   
25478               0               0               0               0   
25479               0               0               0               0   

       case_id_EZYV06  case_id_EZYV07  ...  region_of_employment_South  \
0                   0               0  ...                           0   
1                   0               0  ...                           0   
2                   0               0  ...                           0   
3                   0               0  ...                           0   
4                   0               0  ...                           1   
...               ...             ...  ...                         ...   
25475               0               0  ...                           1   
25476               0               0  ...                           0   
25477               0               0  ...                           1   
25478               0               0  ...                           0   
25479               0               0  ...                           0   

       region_of_employment_West  unit_of_wage_Hour  unit_of_wage_Month  \
0                              1                  1                   0   
1                              0                  0                   0   
2                              1                  0                   0   
3                              1                  0                   0   
4                              0                  0                   0   
...                          ...                ...                 ...   
25475                          0                  0                   0   
25476                          0                  0                   0   
25477                          0                  0                   0   
25478                          1                  0                   0   
25479                          0                  0                   0   

       unit_of_wage_Week  unit_of_wage_Year  full_time_position_N  \
0                      0                  0                     0   
1                      0                  1                     0   
2                      0                  1                     0   
3                      0                  1                     0   
4                      0                  1                     0   
...                  ...                ...                  

- in onhot encoder a new columns will be created

- the new columns equal to number of unique labels

- because of case status two extra columns created

- because of continent 6 extra columns are created

- because of case_id 25480 unique ids, 25480 columns created

In [17]:
count=0
for i in cat_columns:
    count=count+visa_df[i].nunique()
print(count+len(num_columns))

25510


In [18]:
count=0
for i in cat_columns[1:]:
    count=count+visa_df[i].nunique()
print(count)
print(count+len(num_columns))

27
30


In [19]:
for i in cat_columns[1:]:
    print(visa_df[i].unique(),visa_df[i].nunique())

['Asia' 'Africa' 'North America' 'Europe' 'South America' 'Oceania'] 6
['High School' "Master's" "Bachelor's" 'Doctorate'] 4
['N' 'Y'] 2
['N' 'Y'] 2
['West' 'Northeast' 'South' 'Midwest' 'Island'] 5
['Hour' 'Year' 'Week' 'Month'] 4
['Y' 'N'] 2
['Denied' 'Certified'] 2


**Advantages**

- using one hot encoder we will get new columns

- this new column values at a time only one will be **True**, another will be **False**

- which means new columns are **independent each other**

- Independent means **no relation**

- Means no **Multi collinearity**

- Multicollinearity is an important properity in Machine learning

- **Machine learning models expect the input columns should be independent each other**

- So one hot encoder gives the no multi collinearity

**Disadvantages**

- the disadvantage is it will increase many columns

- in the above example the total columns becomes 25510

- this is happened because of case id

- in real time we will drop the id columns

- eventhough if we drop still we are getting 30 columns

- More columns means, more time to process

- we required more infrastructure (CPU-GPU-TPU)

- This is called as **curse of dimensionality**